tensorflow_probability/python/distributions/sratio.py

```diff
@@ -0,0 +1,359 @@
+
+    with tf.name_scope(name) as name:
+
+      float_dtype = dtype_util.common_dtype(
+          [cutpoints, loc],
+          dtype_hint=tf.float32)
+
+      self._cutpoints = tensor_util.convert_nonref_to_tensor(
+          cutpoints, dtype_hint=float_dtype, name='cutpoints')
+      self._loc = tensor_util.convert_nonref_to_tensor(
+          loc, dtype_hint=float_dtype, name='loc')
+
+      super(StoppingRatioLogistic, self).__init__(
+          dtype=dtype,
+          reparameterization_type=reparameterization.NOT_REPARAMETERIZED,
+          validate_args=validate_args,
+          allow_nan_stats=allow_nan_stats,
+          parameters=parameters,
+          name=name)
+
+  @classmethod
+  def _params_event_ndims(cls):
+    return dict(cutpoints=1, loc=0)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return dict(
+        zip(('loc', 'scale'),
+            ([tf.convert_to_tensor(sample_shape, dtype=tf.int32)] * 2)))
+
+  @property
+  def cutpoints(self):
+    """Input argument `cutpoints`."""
+    return self._cutpoints
+
+  @property
+  def loc(self):
+    """Input argument `loc`."""
+    return self._loc
+
+  def categorical_log_probs(self):
+    """Log probabilities for the `K+1` sequential categories."""
+    return tf.math.log(self.categorical_probs())
+
+  def categorical_probs(self):
+    """Probabilities for the `K+1` sequential categories."""
+    cutpoints = self.cutpoints
+    loc = self.loc
+
+    num_cat = self._num_categories()
+    one = tf.ones(1, dtype=cutpoints.dtype)
+
+    q = one - tf.math.sigmoid(cutpoints - loc[..., tf.newaxis])
+    p = one - q
+
+    qs = tf.math.cumprod(q[..., :(num_cat - 2)], axis=-1)
+    p = tf.concat([p[..., :1], p[..., 1:num_cat] * qs], axis=-1)
+    qs = tf.math.reduce_prod(q[..., :num_cat], axis=-1)
+    p = tf.concat([p, qs[..., tf.newaxis]], axis=-1)
+
+    return p
+
+  def _num_categories(self):
+    return prefer_static.shape(self.cutpoints, out_type=self.dtype)[-1] + 1
+
+  def _sample_n(self, n, seed=None):
+    logits = tf.reshape(
+        self.categorical_log_probs(), [-1, self._num_categories()])
+    draws = tf.random.categorical(logits, n, dtype=self.dtype, seed=seed)
+    return tf.reshape(
+        tf.transpose(draws),
+        shape=tf.concat([[n], self._batch_shape_tensor()], axis=0))
+
+  def _batch_shape_tensor(self, cutpoints=None, loc=None):
+    cutpoints = self.cutpoints if cutpoints is None else cutpoints
+    loc = self.loc if loc is None else loc
+    return prefer_static.broadcast_shape(
+        prefer_static.shape(cutpoints)[:-1],
+        prefer_static.shape(loc))
+
+  def _batch_shape(self):
+    return tf.broadcast_static_shape(
+        self.loc.shape, self.cutpoints.shape[:-1])
+
+  def _event_shape_tensor(self):
+    return tf.constant([], dtype=tf.int32)
+
+  def _event_shape(self):
+    return tf.TensorShape([])
+
+  def _log_prob(self, x):
+    num_categories = self._num_categories()
+    x, cat_log_probs = _broadcast_cat_event_and_params(
```

Comment:
Just to make sure, this has to be explicitly broadcasted due to the tf.gather?

Reply:
Yes! Is there another way to do that more efficiently?